strategy_infura.go

```
@@ -0,0 +1,524 @@
+	_, err := as.SetInstanceProtectionWithContext(ctx, &autoscaling.SetInstanceProtectionInput{
+		AutoScalingGroupName: aws.String(asgName),
+		InstanceIds:          ids,
+		ProtectedFromScaleIn: aws.Bool(true),
+	})
+
+	return err
+}
+
+func removeInstanceProtection(ctx context.Context, sess *session.Session, asgName string, instances []instance) error {
+	as := autoscaling.New(sess)
+
+	ids := make([]*string, len(instances))
+	for i, inst := range instances {
+		id := inst.instanceId
+		ids[i] = &id
+	}
+
+	_, err := as.SetInstanceProtectionWithContext(ctx, &autoscaling.SetInstanceProtectionInput{
+		AutoScalingGroupName: aws.String(asgName),
+		InstanceIds:          ids,
+		ProtectedFromScaleIn: aws.Bool(false),
+	})
+
+	return err
+}
+
+// waitForNewInstances:
+// - poll the ASG to detect new instances
+// - wait for them to be ready at the ASG level
+// - if any, wait for them to be ready at the TG level
+// - when ready, immediately enable scale-in protection to prevent them being destroyed
+//
+// Returns when count instance has been found.
+func waitForNewInstances(ctx context.Context, sess *session.Session, asgName string, count int, tg *targetGroup, currentInstances []instance) ([]instance, error) {
+	newInstances, chanErr := detectNewInstances(ctx, sess, asgName, currentInstances)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case err := <-chanErr:
+				if err != nil {
+					log.Printf("\terror while polling ASG details: %v", err)
+				}
+				continue
+			}
+		}
+	}()
+
+	// make sure to stop the following call when done
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	readyInstances, chanErr2 := detectInstancesReady(subCtx, sess, tg, newInstances)
+
+	instanceReady := make([]instance, 0, count)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case err := <-chanErr2:
+			if err != nil {
+				log.Printf("\terror while polling autoscaling or target group reported target health: %v", err)
```

In this case this instance may be caught at the next polling iteration, correct?

---

That's correct. This polling loop only end when the context end. The error is only logged, which means that the next loop can find zero, one or more new instances.

---

To be more clear as it's not so obvious on a github PR: there is **four** endless loop binded the the context lifecycle here:
- one in `detectNewInstances()` doing ASG polling, which output to two **channels** for 1) values and 2) errors
- the goroutine in this function that log and discards the errors from the error channel (line 294)
- the value channel get passed to `detectInstancesReady()` which poll the ASG health check and possibly the TG. This function also output to two channels for values and errors
- this `for` loop that 1) log and discard errors from the second error channel and 2) read from the value channel, set the instance protection and accumulate the healthy instances until we have the expected count